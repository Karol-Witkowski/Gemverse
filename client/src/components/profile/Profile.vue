<template>
  <v-container
    class="mt-16 pt-16"
    fluid
  >
    <v-flex class="mx-auto xs11">
      <v-card
        class="mx-auto"
        max-width="500"
      >
        <v-card-title class="pb-1">
          <span class="font-weight-bold grey--text mx-auto subtitle-1 text--darken-2">
            {{ getUserInfo.username }}
          </span>
        </v-card-title>
        <v-list-item class="mx-auto">
          <v-list-item-avatar class="mx-auto">
            <img
              alt="user avatar"
              :src="`data:image/svg+xml;utf8,${generateAvatar(getUserInfo.username)}`"
            />
          </v-list-item-avatar>
        </v-list-item>
        <v-list-item>
          <v-list-item-content class="mb-0 pb-0">
            <v-list class="d-flex grey--text justify-center subtitle-1 text--darken-2">
              Email address:
              <span class="font-weight-bold pl-1">
                {{ getUserInfo.email }}
              </span>
            </v-list>
          </v-list-item-content>
        </v-list-item>
        <v-list-item>
          <v-list-item-content >
            <v-list class="d-flex grey--text justify-center subtitle-1 text--darken-2">
              Member since:
              <span class="font-weight-bold pl-1">
                {{ showCreationDate(this.getUserInfo.createdDate) }}
              </span>
            </v-list>
          </v-list-item-content>
        </v-list-item>
        <v-list-item>
          <v-list-item-content>
            <v-list-item-action>
                <v-dialog
                  max-width="600"
                  overlay-opacity="0.15"
                  persistent
                  :retain-focus="false"
                  v-model="deleteUserModal"
                >
                  <template
                    class="mb-16"
                    v-slot:activator="{ on, attrs }"
                  >
                    <v-btn
                      class="mb-2 mx-auto"
                      color="secondary"
                      outlined
                      small
                      v-bind="attrs"
                      v-on="on"
                    >
                      delete account*
                    </v-btn>
                  </template>
                  <v-card>
                    <v-card-title class="headline grey--text text--darken-2">
                      Delete user account
                    </v-card-title>
                    <v-card-text>
                      Click "OK" to delete account. Removed account cannot be restored.
                    </v-card-text>
                    <v-card-text
                      class="errorMessage"
                      v-if="deleteError"
                    >
                      {{ deleteError }}
                    </v-card-text>
                    <v-divider />
                    <v-card-actions>
                      <v-btn
                        @click="closeModals"
                        color="primary"
                        outlined
                        text
                      >
                        close
                      </v-btn>
                      <v-spacer />
                      <v-btn
                        @click="deleteUser()"
                        color="primary"
                        outlined
                        text
                      >
                        ok
                      </v-btn>
                    </v-card-actions>
                  </v-card>
                </v-dialog>
              </v-list-item-action>
            <v-list-item class="deleteInfo">
              *Your personal data will be deleted,
              although all messages will be displayed
              (without the user data)
            </v-list-item>
          </v-list-item-content>
        </v-list-item>
      </v-card>
    </v-flex>
  </v-container>
</template>

<script>
import axios from 'axios';
import dayjs from 'dayjs';
import { generateFromString } from 'generate-avatar';
import { mapGetters } from 'vuex';

export default {
  name: 'Profile',
  computed: {
    ...mapGetters(['getUserInfo']),
  },
  data() {
    return {
      deleteError: '',
      deleteRoomModal: false,
      deleteUserModal: false,
      errors: [],
    };
  },

  methods: {
    closeModals() {
      this.deleteError = '';
      this.deleteUserModal = false;
    },

    deleteUser() {
      axios.put('http://localhost:3000/api/user/remove/logged', this.getUserInfo)
        .then(() => {
          this.$store.dispatch('resetState', true);
          localStorage.clear();
          this.$router.push({
            name: 'Login',
            params: { message: 'Account deleted' },
          });
        })
        .catch((error) => {
          console.log(error);
          this.deleteError = error.response.data.error;
        });
    },

    generateAvatar(username) {
      return generateFromString(username);
    },

    showCreationDate(creationDate) {
      return dayjs(creationDate).format('dddd, MMMM D YYYY');
    },
  },
};
</script>

<style lang="scss">
.deleteInfo {
  text-align: justify;
  text-justify: inter-word;
  word-break: normal;
}

.errorMessage {
  color: rgb(194, 57, 57)!important;
}

.userAvatar {
  height: 60px!important;
  width: 60px!important;
}
</style>
