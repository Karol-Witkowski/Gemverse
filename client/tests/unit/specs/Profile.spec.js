import { createLocalVue, mount } from '@vue/test-utils';
import axios from 'axios';
import Vue from 'vue';
import Vuetify from 'vuetify';
import Profile from '@/components/profile/Profile.vue';

const localVue = createLocalVue();
const mockRouter = {
  push: jest.fn(),
};
const url = 'http://localhost:3000/api/user/remove/logged';
let vuetify;
let wrapper;
const error = {
  response: {
    data: {
      error: 'Error msg',
    },
  },
};
const mockStore = {
  dispatch: jest.fn(),
  getters: {
    getUserInfo: {
      createdDate: '2021-04-20T01:01:22.269Z',
      email: 'test@mail.js',
      _id: '321testid',
      username: 'testUser',
    },
  },
};

document.body.setAttribute('data-app', true);
jest.mock('axios');
jest.spyOn(window.localStorage.__proto__, 'clear');
vuetify = new Vuetify();

describe('Implementation test for Profile.vue - successful HTTP delete', () => {
  beforeEach(() => {
    axios.delete.mockResolvedValue();

    wrapper = mount(Profile, {
      localVue,
      mocks: {
        $router: mockRouter,
        $store: mockStore,
      },
      vuetify,
      data() {
        return {
          deleteError: '',
          deleteUserModal: false,
        };
      },
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
    wrapper.destroy();
  });

  it('Render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Initializes with correct elements', () => {
    // Check data binding
    expect(wrapper.findAll('span').at(0).text()).toMatch('testUser');
    expect(wrapper.findAll('span').at(1).text()).toMatch('test@mail.js');
    expect(wrapper.findAll('span').at(2).text()).toMatch('Tuesday, April 20 2021');

    // Check that error message is not visible
    expect(wrapper.find('.errorMessage').exists()).toBeFalsy();
  });

  it('Should sends delete request with correct data on account delete', async () => {
    await wrapper.vm.deleteUser();

    // Check if post was called
    expect(axios.delete).toHaveBeenCalled();

    // Check if post was called once
    expect(axios.delete).toHaveReturnedTimes(1);

    // Check if post was called with correct data
    expect(axios.delete).toHaveBeenCalledWith(
      url,
      {
        _id: '321testid',
        createdDate: '2021-04-20T01:01:22.269Z',
        email: 'test@mail.js',
        username: 'testUser',
      },
    );
  });

  it('Should call dispatch method on delete account', async () => {
    await wrapper.vm.deleteUser();

    // Check if any action were dispatched
    expect(mockStore.dispatch).toHaveBeenCalled();

    // Check if one action was dispatched
    expect(mockStore.dispatch).toHaveReturnedTimes(1);

    // Check if action was dispatched with correct data
    expect(mockStore.dispatch).toHaveBeenCalledWith(
      'resetState',
      true,
    );
  });

  it('Should call store clear method on delete account', async () => {
    await wrapper.vm.deleteUser();

    // Check if local storage clear method was called
    expect(localStorage.clear).toHaveBeenCalled();

    // Check if local storage clear method was called once
    expect(localStorage.clear).toHaveReturnedTimes(1);
  });

  it('Redirect user to login page on account delete', async () => {
    await wrapper.vm.deleteUser();

    // Check if router push method was called
    expect(mockRouter.push).toHaveBeenCalled();

    // Check if router push method was called one time
    expect(mockRouter.push).toHaveBeenCalledTimes(1);

    // Check if router push method was called with correct object
    expect(mockRouter.push).toHaveBeenCalledWith({
      name: 'Login',
      params: {
        message: 'Account deleted',
      },
    });
  });
});

describe('Implementation test for Profile.vue - failed HTTP delete', () => {
  beforeEach(() => {
    axios.delete.mockRejectedValue(error);

    wrapper = mount(Profile, {
      localVue,
      mocks: {
        $router: mockRouter,
        $store: mockStore,
      },
      vuetify,
      data() {
        return {
          deleteError: '',
          deleteUserModal: false,
        };
      },
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
    wrapper.destroy();
  });

  it('Should display error on account delete failure', async () => {
    wrapper.setData({
      deleteUserModal: true,
    });

    await wrapper.vm.deleteUser();

    await Vue.nextTick();

    expect(wrapper.vm.deleteError).toBe('Error msg');
    expect(wrapper.find('.errorMessage').exists()).toBeTruthy();
    expect(wrapper.findAll('.errorMessage').at(0).text()).toBe('Error msg');
  });

  it('Should clear errors and close dialog on canceled account delete', async () => {
    wrapper.setData({
      deleteUserModal: true,
    });

    wrapper.vm.deleteUser();

    await Vue.nextTick();

    wrapper.vm.closeModal();

    expect(wrapper.vm.deleteError).toMatch('');
    expect(wrapper.vm.deleteUserModal).toBeFalsy();
  });
});

describe('Behavioral test for Profile.vue - successful HTTP delete', () => {
  beforeEach(() => {
    axios.delete.mockResolvedValue();

    wrapper = mount(Profile, {
      localVue,
      mocks: {
        $router: mockRouter,
        $store: mockStore,
      },
      vuetify,
      data() {
        return {
          deleteError: '',
          deleteUserModal: false,
        };
      },
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
    wrapper.destroy();
  });

  it('Should sends delete request with correct data on delete button click', async () => {
    await wrapper.find('button[name=delete]').trigger('click');

    // Check if click event opened dialog
    expect(wrapper.vm.deleteUserModal).toBeTruthy();

    // Confirm account delete
    await wrapper.find('button[name=accept]').trigger('click');

    // Check if post was called;
    expect(axios.delete).toHaveBeenCalled();

    // Check if post was called once
    expect(axios.delete).toHaveReturnedTimes(1);

    // Check if post was called with correct data
    expect(axios.delete).toHaveBeenCalledWith(
      url,
      {
        _id: '321testid',
        createdDate: '2021-04-20T01:01:22.269Z',
        email: 'test@mail.js',
        username: 'testUser',
      },
    );
  });

  it('Should call dispatch method on delete button click', async () => {
    await wrapper.find('button[name=delete]').trigger('click');

    // Check if click event opened dialog
    expect(wrapper.vm.deleteUserModal).toBeTruthy();

    // Confirm account delete
    await wrapper.find('button[name=accept]').trigger('click');

    // Check if any action were dispatched
    expect(mockStore.dispatch).toHaveBeenCalled();

    // Check if one action was dispatched
    expect(mockStore.dispatch).toHaveReturnedTimes(1);

    // Check if action was dispatched with correct data
    expect(mockStore.dispatch).toHaveBeenCalledWith(
      'resetState',
      true,
    );
  });

  it('Should call store clear method on delete button click', async () => {
    await wrapper.find('button[name=delete]').trigger('click');

    // Check if click event opened dialog
    expect(wrapper.vm.deleteUserModal).toBeTruthy();

    // Confirm account delete
    await wrapper.find('button[name=accept]').trigger('click');

    // Check if local storage clear method was called
    expect(localStorage.clear).toHaveBeenCalled();

    // Check if local storage clear method was called once
    expect(localStorage.clear).toHaveReturnedTimes(1);
  });

  it('Redirect user to login page on delete button click', async () => {
    await wrapper.find('button[name=delete]').trigger('click');

    // Check if click event opened dialog
    expect(wrapper.vm.deleteUserModal).toBeTruthy();

    // Confirm account delete
    await wrapper.find('button[name=accept]').trigger('click');

    // Check if router push was called
    expect(mockRouter.push).toHaveBeenCalled();

    // Check if router push was called one time
    expect(mockRouter.push).toHaveBeenCalledTimes(1);

    // Check if router push was called with correct object
    expect(mockRouter.push).toHaveBeenCalledWith({
      name: 'Login',
      params: {
        message: 'Account deleted',
      },
    });
  });
});

describe('Behavioral test for Profile.vue - failed HTTP delete', () => {
  beforeEach(() => {
    axios.delete.mockRejectedValue(error);

    wrapper = mount(Profile, {
      localVue,
      mocks: {
        $router: mockRouter,
        $store: mockStore,
      },
      vuetify,
      data() {
        return {
          deleteError: '',
          deleteUserModal: false,
        };
      },
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
    wrapper.destroy();
  });

  it('Should display error on delete button click', async () => {
    await wrapper.find('button[name=delete]').trigger('click');

    // Check if click event opened dialog
    expect(wrapper.vm.deleteUserModal).toBeTruthy();

    // Confirm account delete
    await wrapper.find('button[name=accept]').trigger('click');

    await Vue.nextTick();

    expect(wrapper.find('.errorMessage').exists()).toBeTruthy();
    expect(wrapper.findAll('.errorMessage').at(0).text()).toBe('Error msg');
  });

  it('Should clear errors and close dialog on delete cancel button click', async () => {
    await wrapper.find('button[name=delete]').trigger('click');

    // Check if click event opened dialog
    expect(wrapper.vm.deleteUserModal).toBeTruthy();

    // Confirm account delete
    await wrapper.find('button[name=accept]').trigger('click');

    await Vue.nextTick();

    // Check if errors are visible
    expect(wrapper.find('.errorMessage').exists()).toBeTruthy();
    expect(wrapper.findAll('.errorMessage').at(0).text()).toBe('Error msg');

    // Close dialog
    await wrapper.find('button[name=close]').trigger('click');

    // Check if data returned to the initial state
    expect(wrapper.vm.deleteError).toMatch('');
    expect(wrapper.vm.deleteUserModal).toBeFalsy();
  });
});
