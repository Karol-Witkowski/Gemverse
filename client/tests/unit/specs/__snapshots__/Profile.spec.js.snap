// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Implementation test for Profile.vue Render correctly 1`] = `
<div class="container mt-16 pt-16 container--fluid">
  <div class="flex mx-auto xs11">
    <div class="mx-auto v-card v-sheet theme--light" style="max-width: 500px;">
      <div class="v-card__title pb-1"><span class="font-weight-bold grey--text mx-auto subtitle-1 text--darken-2">
          testUser
        </span></div>
      <div tabindex="-1" class="mx-auto v-list-item theme--light">
        <div class="v-avatar v-list-item__avatar mx-auto" style="height: 40px; min-width: 40px; width: 40px;"><img alt="user avatar" src="data:image/svg+xml;utf8,<svg width=&quot;300&quot; height=&quot;300&quot; viewBox=&quot;0 0 300 300&quot; xmlns=&quot;http://www.w3.org/2000/svg&quot;><rect id=&quot;bg&quot; width=&quot;300&quot; height=&quot;300&quot; fill=&quot;rgb(51,239,55)&quot; /><path d=&quot;m 150 536 Q 81 219 -236 150 Q 81 81 150 -236 Q 219 81 536 150 Q 219 219 150 536 z&quot; fill=&quot;rgb(84,158,159)&quot; /><path d=&quot;m 150 462 Q 57 243 -162 150 Q 57 57 150 -162 Q 243 57 462 150 Q 243 243 150 462 z&quot; fill=&quot;rgb(171,97,96)&quot; /><path d=&quot;m 150 281 Q 173 127 19 150 Q 173 173 150 19 Q 127 173 281 150 Q 127 127 150 281 z&quot; fill=&quot;rgb(204,16,200)&quot; /></svg>"></div>
      </div>
      <div tabindex="-1" class="v-list-item theme--light">
        <div class="v-list-item__content mb-0 pb-0">
          <div role="list" class="v-list d-flex grey--text justify-center subtitle-1 text--darken-2 v-sheet theme--light">
            Email address:
            <span class="font-weight-bold pl-1">
              test@mail.js
            </span>
          </div>
        </div>
      </div>
      <div tabindex="-1" class="v-list-item theme--light">
        <div class="v-list-item__content">
          <div role="list" class="v-list d-flex grey--text justify-center subtitle-1 text--darken-2 v-sheet theme--light">
            Member since:
            <span class="font-weight-bold pl-1">
              Tuesday, April 20 2021
            </span>
          </div>
        </div>
      </div>
      <div tabindex="-1" class="v-list-item theme--light">
        <div class="v-list-item__content">
          <div class="v-list-item__action"><button type="button" class="mb-2 mx-auto v-btn v-btn--outlined theme--light v-size--small secondary--text" role="button" aria-haspopup="true" aria-expanded="false"><span class="v-btn__content">
                    delete account*
                  </span></button>
            <div role="dialog" class="v-dialog__container">
              <!---->
            </div>
          </div>
          <div tabindex="-1" class="deleteInfo v-list-item theme--light">
            *Your personal data will be deleted,
            although all messages will be displayed
            (without the user data)
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
